Python/lock.c

```diff
@@ -584,3 +570,19 @@ uint32_t _PySeqLock_AfterFork(_PySeqLock *seqlock)
 
      return 0;
 }
+
+#undef PyMutex_Lock
+void
+PyMutex_Lock(PyMutex *m)
+{
+    _PyMutex_LockTimed(m, -1, _PY_LOCK_DETACH);
+}
+
+#undef PyMutex_Unlock
+void
+PyMutex_Unlock(PyMutex *m)
+{
+    if (_PyMutex_TryUnlock(m) < 0) {
+        Py_FatalError("unlocking mutex that is not locked");
```

Comment:
I suggest to only issue a fatal error if Python is built in debug mode or in Python Development Mode (-X dev). Otherwise, emit a warning. If the warning raises an exception (-Werror), log the warning as an unraisable exception.

```suggestion
        int fatal_error;
#ifdef Py_DEBUG
        fatal_error = 1;
#else
        fatal_error = _Py_GetConfig()->dev_mode;
#endif
        if (fatal_error) {
            Py_FatalError("unlocking mutex that is not locked");
        }
        else {
            if (PyErr_WarnEx(PyExc_SystemError,
                    "unlocking mutex that is not locked", 1))
            {
                PyErr_WriteUnraisable(NULL);
            }
        }
```

Reply:
Mutexes are too low level to handle this way. We voted against requiring a return value to indicate failure, which means the only viable option is to tear down the entire process with a fatal error.

Reply:
Yeah, it's not safe to issue a warning here because the you might not have a valid thread state / hold the GIL. I don't think it's worth trying to do something more clever than `Py_FatalError` here because you run the risk of your error reporting mechanism crashing and hiding the original programmer error.

Reply:
Oh ok, that's sad, but I get it :-)